Release v0.2 TBD (#1090)
hyperledger/solang: CHANGELOG.md

All notable changes to [Solang](https://github.com/hyperledger/solang/)
will be documented here.

## v0.2.0.0 Berlin

> We are happy to release solang `v0.2.0` codename `Berlin` today. Aside from containing many smaller fixes and improvements, this release marks a milestone towards maturing our Substrate compilation target: Any regressions building up since `ink!` v3.0 are fixed, most notably the metadata format (shoutout and many thanks to external contributor [extraymond](https://github.com/extraymond)) and event topics. Furthermore, we are leaving `ink!` version 3 behind us, in favor of introducing compatibility with the recent `ink!` 4 beta release and the latest pallet contract `v0.22.1`.

Reviewer: ```suggestion
We are happy to release solang `v0.2.0` codenamed `Berlin` today. Aside from containing many small fixes and improvements, this release marks a milestone towards maturing our Substrate compilation target: Any regressions building up since `ink!` v3.0 are fixed, most notably the metadata format (shoutout and many thanks to external contributor [extraymond](https://github.com/extraymond)) and event topics. Furthermore, we are leaving `ink!` version 3 behind us, in favor of introducing compatibility with the recent `ink!` 4 beta release and the latest pallet contract `v0.22.1`.
```

> We are happy to release solang `v0.2.0` codename `Berlin` today. Next to containing many smaller fixes and improvements, this release marks a milestone towards maturing our substrate compilation target: Any regressions building up since `ink!` v3.0 are fixed, most notably the metadata format (shoutout and many thanks to external contributor [extraymond](https://github.com/extraymond)) and event topics. Furthermore, we are leaving `ink!` version 3 behind us already, in favor of introducing compatibility with the recent `ink!` 4 beta release and the latest pallet contract `v0.22.1`.

Reviewer: ```suggestion
We are happy to release solang `v0.2.0` codename `Berlin` today. Aside from containing many smaller fixes and improvements, this release marks a milestone towards maturing our Substrate compilation target: Any regressions building up since `ink!` v3.0 are fixed, most notably the metadata format (shoutout and many thanks to external contributor [extraymond](https://github.com/extraymond)) and event topics. Furthermore, we are leaving `ink!` version 3 behind us, in favor of introducing compatibility with the recent `ink!` 4 beta release and the latest pallet contract `v0.22.1`.
```

`Next to` usually refers to physical proximity. Check [this](https://english.stackexchange.com/questions/242823/is-next-to-that-really-an-alternative-to-additionally-or-moreover) for clarification.

### Added
- **Solana / breaking:** The try-catch construct is no longer permitted on Solana, as it
   never worked. Any CPI error will abort the transaction.
   [seanyoung](https://github.com/seanyoung)
- **Solana:** Introduce new sub-command `solang idl` which can be used for generating
  a Solidity interface file from an Anchor IDL file. This can be used for calling
  Anchor Contracts on Solana. [seanyoung](https://github.com/seanyoung)
- **Substrate:** Provide specific Substrate builtins via a "substrate" file. The

>   `Hash` type from `ink!` is the first type made available for Solidity contracts.

Reviewer: `is the first type made available for Solidity contracts` 
This needs rephrasing. Is it the first Substrate specific type?

Author: Exactly

>   `Hash` type from `ink!` is the first type made available for solidity contracts.

Reviewer: ```suggestion
  `Hash` type from `ink!` is the first type made available for Solidity contracts.
```
The first Substrate specific type?

  [xermicus](https://github.com/xermicus)
- **Substrate:** Introduce the `--log-api-return-codes` CLI flag, which changes the
  emitted code to print return codes for `seal` API calls into the debug buffer.
  [xermicus](https://github.com/xermicus)
- Introduce function name mangling for overloaded functions and constructors, so
  that they can be represented properly in the metadata.
  [xermicus](https://github.com/xermicus)

### Changed
 - The Solana target now uses Borsh encoding rather than eth abi

>    encoding. This is towards making Solang contracts Anchor compatible.

Reviewer: ```suggestion
   encoding. This is aimed at making Solang contracts Anchor compatible.
```

Reviewer: I believe this sounds better.

   [LucasSte](https://github.com/LucasSte)
- **Substrate / breaking:** Supported node version is now pallet contracts `v0.22.1`.
  [xermicus](https://github.com/xermicus)
- **Substrate / breaking:** Remove the deprecated `random` builtin.
  [xermicus](https://github.com/xermicus)

### Fixed

>  - The parser does not give up after the first error, when possible.

Reviewer: ```suggestion
 - Whenever possible, the parser does not give up after the first error.
```
The other version sounds awkward.

   [salaheldinsoliman](https://github.com/salaheldinsoliman)
 - Constant expressions are checked for overflow.
   [salaheldinsoliman](https://github.com/salaheldinsoliman)

>  - AddMod and MulMod were broken, this is now fixed.

Reviewer: ```suggestion
 - AddMod and MulMod were broken. This is now fixed.
```
This was a comma splice.

   [LucasSte](https://github.com/LucasSte)
- **Substrate / breaking:** Solang is now compatible with `ink!` version 4 (beta).
  [xermicus](https://github.com/xermicus)
- **Substrate:** Switched ABI generation to use official `ink!` crates, which fixes all
  remaining metadata regressions.
  [extraymond](https://github.com/extraymond) and [xermicus](https://github.com/xermicus)
- **Substrate:** Allow constructors to have a name, so that multiple constructors are 
  supported, like in `ink!`.
  [xermicus](https://github.com/xermicus)
- Provided examples and most of the solidity code snippets in our documentation

>   are now checked for succesful compilation on the solang CI.

Reviewer: ```suggestion
  are now checked for succesful compilation on the Solang CI.
```

> - Provided examples and most of the solidity code snippets in our documentation
>   are now checked for succesful compilation on the Solang CI.
>   [xermicus](https://github.com/xermicus)

Reviewer: ```suggestion
- Provided examples and checked most of the solidity code snippets in our documentation
  for succesful compilation on the Solang CI.
  [xermicus](https://github.com/xermicus)
```
There needs to be parallelism between the verbal usage after and before the `and`.

Author: Sorry I don't get this one. It says that both, the "provided examples" and "most of the solidity code snippets in our documentation" are check for succesfull compilation? What is wrong with this sentence?

Author: @LucasSte I rephrased that sentence, WDYT?

Reviewer: > Sorry I don't get this one. It says that both, the "provided examples" and "most of the solidity code snippets in our documentation" are checked for successful compilation? What is wrong with this sentence?

Before the `and` you wrote `provided` without a subject. After the `and` you wrote `are checked` (in passive voice) with a subject. Whatever construction you choose, it must remain unchanged before and after the `and`. If you choose to use active voice and a subject, this construction must be the same for both clauses (before and after the preposition).

Reviewer: > @LucasSte I rephrased that sentence, WDYT?

Sounds better now.

> - **Substrate:** Fix events with topics. The topic hashes generated by solang

Reviewer: ```suggestion
- **Substrate:** Fix events with topics. The topic hashes generated by Solang
```

All notable changes to [Solang](https://github.com/hyperledger/solang/)
will be documented here.

## v0.2.0.0 Berlin

> We are happy to release solang `v0.2.0` codenamed `Berlin` today. Aside from containing many small fixes and improvements, this release marks a milestone towards maturing our Substrate compilation target: Any regressions building up since `ink!` v3.0 are fixed, most notably the metadata format (shoutout and many thanks to external contributor [extraymond](https://github.com/extraymond)) and event topics. Furthermore, we are leaving `ink!` version 3 behind us, in favor of introducing compatibility with the recent `ink!` 4 beta release and the latest pallet contract `v0.22.1`.

Reviewer: ```suggestion
We are happy to release solang `v0.2.0` codenamed `Berlin` today. Aside from containing many small fixes and improvements, this release marks a milestone towards maturing our Substrate compilation target: any regressions building up since `ink!` v3.0 are fixed, most notably the metadata format (shoutout and many thanks to external contributor [extraymond](https://github.com/extraymond)) and event topics. Furthermore, we are leaving `ink!` version 3 behind us, in favor of introducing compatibility with the recent `ink!` 4 beta release and the latest pallet contract `v0.22.1`.
```

   [LucasSte](https://github.com/LucasSte)
- **Substrate / breaking:** Supported node version is now pallet contracts `v0.22.1`.
  [xermicus](https://github.com/xermicus)
- **Substrate / breaking:** Remove the deprecated `random` builtin.
  [xermicus](https://github.com/xermicus)

### Fixed
 - Whenever possible, the parser does not give up after the first error.
   [salaheldinsoliman](https://github.com/salaheldinsoliman)
 - Constant expressions are checked for overflow.
   [salaheldinsoliman](https://github.com/salaheldinsoliman)
 - AddMod and MulMod were broken. This is now fixed.
   [LucasSte](https://github.com/LucasSte)
- **Substrate / breaking:** Solang is now compatible with `ink!` version 4 (beta).
  [xermicus](https://github.com/xermicus)
- **Substrate:** Switched ABI generation to use official `ink!` crates, which fixes all
  remaining metadata regressions.
  [extraymond](https://github.com/extraymond) and [xermicus](https://github.com/xermicus)
- **Substrate:** Allow constructors to have a name, so that multiple constructors are 
  supported, like in `ink!`.
  [xermicus](https://github.com/xermicus)

> - All provided examples as well as most of the solidity code snippets in our

Reviewer: ```suggestion
- All provided examples as well as most of the Solidity code snippets in our
```

## Unreleased

> ## v0.2.0 TBD

Reviewer 2: Nitpick: no need to have `## Unreleased` in the release

> 

Reviewer 2: I think this a good point to have a paragraph about major changes, like the improved substrate compatibility.

Author: ACK, I'm gonna write something

  [xermicus](https://github.com/xermicus)
- Introduce function name mangling for overloaded functions and constructors, so
  that they can be represented properly in the metadata.
  [xermicus](https://github.com/xermicus)
- Substrate: Remove the depricated `random` builtin.
  [xermicus](https://github.com/xermicus)

### Fixed
- Substrate: Solang is now compatible with `ink!` version 4 (beta).
  [xermicus](https://github.com/xermicus)
- Substrate: Switched ABI generation to use official `ink!` crates, which fixes all
  remaining metadata regressions.
  [extraymond](https://github.com/extraymond) and [xermicus](https://github.com/xermicus)
- Substrate: Allow constructors to have a name, so that multiple constructors are 
  supported, like in `ink!`.
  [xermicus](https://github.com/xermicus)
- Provided examples and most of the solidity code snippets in our documentation
  are now checked for succesful compilation on the solang CI.
  [xermicus](https://github.com/xermicus)
- Substrate: Fix events with topics. The topic hashes generated by solang
  contracts are now exactly the same as those generated by `ink!`.
  [xermicus](https://github.com/xermicus)

> 

Reviewer 2: @xermicus can you please add this:
```
### Fixed

 - The parser does not give up after the first error, when possible.
   [salaheldinsoliman](https://github.com/salaheldinsoliman)
 - Constant expressions are checked for overflow.
   [salaheldinsoliman](https://github.com/salaheldinsoliman)
 - AddMod and MulMod were broken, this is now fixed.
   [LucasSte](https://github.com/LucasSte)

### Changed
 - The Solana target now uses Borsh encoding rather than eth abi
   encoding. This is towards making Solang contracts Anchor compatible.
   [LucasSte](https://github.com/LucasSte)

### Removed
 - The try-catch construct is no longer permitted on Solana, as it
   never worked. Any CPI error will abort the transaction.
   [seanyoung](https://github.com/seanyoung)
```
